File: net/JetBrains.Serialization/src/Dataholders/AttributeInfos/MSCertExtensionsAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;
using JetBrains.Serialization;

[JsonObject(MemberSerialization.OptIn)]
public class MSCertExtensionsAttributeInfo : AttributeInfo
{
  [JsonProperty("Identifier")] protected override TextualInfo Identifier { get; }

  [JsonProperty("Content")] private List<List<TextualInfo>> _content;

  [JsonConstructor]
  public MSCertExtensionsAttributeInfo(TextualInfo identifier, List<List<TextualInfo>> content)
  {
    Identifier = identifier;
    _content = content;
  }

  public MSCertExtensionsAttributeInfo(Attribute attribute)
  {
    Identifier = TextualInfo.GetInstance(attribute.AttrType);
    _content = attribute.AttrValues.ToArray().OfType<Asn1Sequence>().Select(seq =>
      seq.OfType<Asn1Encodable>().Select(TextualInfo.GetInstance).ToList()).ToList();
  }

  public override Asn1Encodable GetPrimitiveContent() =>
    _content.Select(list => list.ToPrimitiveDerSequence()).ToDerSet();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/attributeInfos/MSCertExtensionsAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSet
import com.jetbrains.signatureverifier.serialization.toPrimitiveDLSequence
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.cms.Attribute

// 1.3.6.1.4.1.311.2.1.11
@Serializable
data class MSCertExtensionsAttributeInfo(
  override val identifier: TextualInfo,
  val content: List<List<TextualInfo>>
) : AttributeInfo() {

  constructor(attribute: Attribute) : this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attributeValues.map { (it as DLSequence).map { s -> TextualInfo.getInstance(s) } }
  )

  override fun getPrimitiveContent() = content.map {
    it.toPrimitiveDLSequence()
  }.toDLSet()
}